### PR TITLE
[SPARK-14821][SQL] Implement AnalyzeTable in sql/core and remove HiveSqlAstBuilder

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

This patch moves analyze table parsing into SparkSqlAstBuilder and removes HiveSqlAstBuilder.

In order to avoid extensive refactoring, I created a common trait for CatalogRelation and MetastoreRelation, and match on that. In the future we should probably just consolidate the two into a single thing so we don't need this common trait.
## How was this patch tested?

Updated unit tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTable.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+
+/**
+ * Analyzes the given table in the current database to generate statistics, which will be
+ * used in query optimizations.
+ *
+ * Right now, it only supports Hive tables and it only updates the size of a Hive table
+ * in the Hive metastore.
+ */
+case class AnalyzeTable(tableName: String) extends RunnableCommand {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTable.scala]: this is mostly moved from the other file

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTable.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        val catalogTable: CatalogTable = relation.catalogTable
+        // This method is mainly based on
+        // org.apache.hadoop.hive.ql.stats.StatsUtils.getFileSizeForTable(HiveConf, Table)
+        // in Hive 0.13 (except that we do not use fs.getContentSummary).
+        // TODO: Generalize statistics collection.
+        // TODO: Why fs.getContentSummary returns wrong size on Jenkins?
+        // Can we use fs.getContentSummary in future?
+        // Seems fs.getContentSummary returns wrong table size on Jenkins. So we use
+        // countFileSize to count the table size.
+        val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTable.scala]: i changed this to use a sql conf rather than hive conf

**Comment by yhuai** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTable.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+          }.getOrElse(0L)
+
+        // Update the Hive metastore if the total size of the table is different than the size
+        // recorded in the Hive metastore.
+        // This logic is based on org.apache.hadoop.hive.ql.exec.StatsTask.aggregateStats().
+        if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
+          sessionState.catalog.alterTable(
+            catalogTable.copy(
+              properties = relation.catalogTable.properties +
+                (AnalyzeTable.TOTAL_SIZE_FIELD -> newTotalSize.toString)))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTable.scala]: this was using some hive constant -- changed to use an internal constant.

**Comment by yhuai** [comment]
LGTM

**Comment by andrewor14** [comment]
LGTM2

**Comment by rxin** [comment]
Merging in master.

**Pull request closed without merging**